File: Carrier.Swift.SDK/ElastosCarrier/Carrier/Group.swift
/*
 * Copyright (c) 2018 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import Foundation

/// The class representing carrier group.
@objc(ELACarrierGroup)
public class CarrierGroup: NSObject {
    /// Carrier group message max length.
    public static let MAX_APP_MESSAGE_LEN: Int = 2048

    private static let TAG: String = "CarrierGroup"
    private static let MAX_GROUP_TITLE_LEN: Int = 127
    private static let MAX_ADDRESS_LEN: Int = 52
    private static let MAX_ID_LEN: Int = 45

    internal var ccarrier: OpaquePointer?
    internal var groupid : String
    private  var didLeave : Bool
    weak var delegate: CarrierGroupDelegate?

    init(_ ccarrier: OpaquePointer!, _ groupid: String,
                 _ delegate: CarrierGroupDelegate) {

        self.ccarrier = ccarrier
        self.groupid  = groupid
        self.delegate = delegate
        self.didLeave = false

        super.init()
    }

    deinit {
        leave()
    }

    func leave() {
        objc_sync_enter(self)

        if (!didLeave) {
            Log.d(CarrierGroup.TAG, "Destroying group \(groupid) ...");

            self.ccarrier = nil
            self.delegate = nil
            self.didLeave = true
            Log.i(CarrierGroup.TAG, "Group \(groupid) destroyed")
        }

        objc_sync_exit(self)
    }

    /// Get groupid in string way.
    ///
    /// - Returns:
    ///     The groupid in string.
    ///
    public func getId() -> String {
        return self.groupid
    }

    /// Invite a specified friend into group.
    ///
    /// - Parameters:
    ///     - friendId: The friend wll be invited into group.
    ///
    /// - Throws:
    ///     CarrierError.
    ///
    @objc(inviteFriend:error:)
    public func inviteFriend(_ friendId: String) throws {
        let result = groupid.withCString{ (cgroupid) -> Int32 in
            return friendId.withCString { (cfriendid) -> Int32 in
                return ela_group_invite(ccarrier, cgroupid, cfriendid)
            }
        }

        guard result >= 0 else {
            let errno: Int = getErrorCode()
            Log.e(CarrierGroup.TAG, "Invite friend \(friendId) error: 0x%X", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }

        Log.d(CarrierGroup.TAG, "Invite friend \(friendId) to group \(groupid) success.")
    }

    /// Send a message to a group.
    ///
    /// The message length may not exceed MAX_APP_MESSAGE_LEN. Larger messages
    /// must be split by application and sent as separate fragments. Other carrier
    /// nodes can reassemble the fragments.
    ///
    /// - Parameters:
    ///   - data: The data content to send.
    ///
    /// - Throws:
    ///     CarrierError
    ///
    @objc(sendMessage:error:)
    public func sendMessage(_ data: Data) throws {
        let result = groupid.withCString{ (cgroupid) -> Int32 in
            return data.withUnsafeBytes{ (cdata) -> Int32 in
                return ela_group_send_message(ccarrier, cgroupid, cdata, data.count)
            }
        }

        guard result >= 0 else {
            let errno: Int = getErrorCode()
            Log.e(CarrierGroup.TAG, "Send group message to \(groupid) error: 0x%X", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }

        Log.d(CarrierGroup.TAG, "Sended group \(groupid) message: \(data).")
    }

    /// Get group title.
    ///
    /// - Returns:
    ///     The new group title.
    ///
    /// - Throws:
    ///     CarrierError
    ///
    @objc(getTitle:)
    public func getTitle() throws -> String {
        let len = CarrierGroup.MAX_GROUP_TITLE_LEN + 1
        var data = Data(count: len)

        let title = groupid.withCString{ (cgroupid) -> String? in
            return data.withUnsafeMutableBytes() {
                (ptr: UnsafeMutablePointer<Int8>) -> String? in
                let result = ela_group_get_title(ccarrier, cgroupid, ptr, len)
                return result >= 0 ? String(cString: ptr) : nil
            }
        }

        guard title != nil else {
            let errno: Int = getErrorCode();
            Log.e(CarrierGroup.TAG, "Get group \(groupid) title error: 0x%X", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }

        Log.d(CarrierGroup.TAG, "Current group \(groupid) title: \(title!)")
        return title!
    }

    /// Set new group title.
    ///
    /// - Parameters:
    ///     - newTitle: new group title
    ///
    /// - Throws:
    ///     CarrierError
    ///
    @objc(setTitle:error:)
    public func setTitle(_ newTitle: String) throws {
        let result = groupid.withCString{ (cgroupid) -> Int32 in
            return newTitle.withCString{ (ctitle) -> Int32 in
                return ela_group_set_title(ccarrier, cgroupid, ctitle)
            }
        }

        guard result >= 0 else {
            let errno: Int = getErrorCode()
            Log.e(CarrierGroup.TAG, "Set group \(groupid) title error: 0x%x", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }
    }

    /// Get group peer list
    ///
    /// - Returns:
    ///     The array of CarrierGroupPeer instance.
    ///
    /// - Throws:
    ///     CarrierError
    ///
    @objc(getPeers:)
    public func getPeers() throws -> [CarrierGroupPeer] {
        let cb: CGroupPeersIterateCallback = { (cpeer, ctxt) in
            if cpeer != nil {
                let cGroupPeer = cpeer!.assumingMemoryBound(to: CGroupPeer.self).pointee
                let peer = convertCGroupPeerToCarrierGroupPeer(cGroupPeer)
                ctxt!.assumingMemoryBound(to: [CarrierGroupPeer].self)
                    .pointee.append(peer)
            }
            return true
        }

        var peers = [CarrierGroupPeer]()
        let result = withUnsafeMutablePointer(to: &peers) { (ptr) -> Int32 in
            return groupid.withCString{ (cgroupid) -> Int32 in
                let cctxt = UnsafeMutablePointer(ptr)
                return ela_group_get_peers(ccarrier, cgroupid, cb, cctxt)
            }
        }

        guard result >= 0 else {
            let errno: Int = getErrorCode()
            Log.e(CarrierGroup.TAG, "Get group \(groupid) peers error: 0x%X", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }

        Log.d(CarrierGroup.TAG, "Current group peers listed below: +++>>")
        for peer in peers {
            Log.d(CarrierGroup.TAG, "\(peer)");
        }
        Log.d(CarrierGroup.TAG, "<<+++")

        return peers
    }

    /// Get group peer
    ///
    /// - Returns:
    ///     The CarrierGroupPeer instance.
    ///
    /// - Throws:
    ///     CarrierError
    ///
    @objc(getPeerByPeerId:error:)
    public func getPeer(byPeerid peerId: String) throws -> CarrierGroupPeer {
        var cpeer = CGroupPeer()
        
        let result = peerId.withCString { (cpeerid) -> Int32 in
            return groupid.withCString{ (cgroupid) -> Int32 in
                return ela_group_get_peer(ccarrier, cgroupid, cpeerid, &cpeer)
            }
        }

        guard result >= 0 else {
            let errno: Int = getErrorCode()
            Log.e(CarrierGroup.TAG, "Get peer \(peerId) info error: 0x%X", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }

        let peer = convertCGroupPeerToCarrierGroupPeer(cpeer)
        Log.d(CarrierGroup.TAG, "Current peer infos: \(peer)")
        return peer
    }
}

File: Carrier.Android.SDK/app/src/main/java/org/elastos/carrier/Group.java
/*
 * Copyright (c) 2018 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.carrier;

import java.util.List;
import java.util.ArrayList;

import org.elastos.carrier.exceptions.CarrierException;

public class Group {
	/**
	 * Carrier App message max length.
	 */
	public static final int MAX_APP_MESSAGE_LEN = 1024;

	/**
	 * Max Carrier group title length.
	 */
	public static final int MAX_GROUP_TITLE_LEN = 127;

	private String groupId;
	private Carrier carrier;

	private static native String new_group(Carrier carrier);
	private static native String group_join(Carrier carrier, String friendId, byte[] cookie);
	private native boolean leave_group(Carrier carrier, String groupId);
	private native boolean group_invite(Carrier carrier, String groupId, String friendId);

	private native boolean group_send_message(Carrier carrier, String groupId, byte[] message);
	private native String group_get_title(Carrier carrier, String groupId);
	private native boolean group_set_title(Carrier carrier, String groupId, String title);
	private native boolean group_get_peers(Carrier carrier, String groupId, GroupPeersIterator iterator, Object context);
	private native PeerInfo group_get_peer(Carrier carrier, String groupId, String peerId);

	private static native int get_error_code();

	Group(Carrier carrier, String groupId) {
		this.carrier = carrier;
		this.groupId = groupId;
	}

	Group(Carrier carrier) throws CarrierException {
		groupId = new_group(carrier);
		if (groupId == null)
			throw CarrierException.fromErrorCode(get_error_code());

		this.carrier = carrier;
	}

	Group(Carrier carrier, String friendId, byte[] cookie) throws CarrierException {
		groupId = group_join(carrier, friendId, cookie);
		if (groupId == null)
			throw CarrierException.fromErrorCode(get_error_code());

		this.carrier = carrier;
	}

	void leave() throws CarrierException {
		if (!leave_group(carrier, groupId))
			throw CarrierException.fromErrorCode(get_error_code());
	}

	String getId() {
		return groupId;
	}

	/**
	 * Invite a specified friend into group.
	 *
	 * @param
	 *	  friendId	The invited friendId
	 *
	 * @throws IllegalArgumentException illegal exception.
	 * @throws CarrierException  carrier exception.
	 */
	public void invite(String friendId) throws CarrierException {
		if (friendId == null || friendId.length() == 0)
			throw new IllegalArgumentException();

		if (!group_invite(carrier, groupId, friendId))
			throw CarrierException.fromErrorCode(get_error_code());
	}

	/**
	 * Send a message to a group.
	 *
	 * The message length may not exceed MAX_APP_MESSAGE_LEN. Larger messages
	 * must be split by application and sent as separate fragments. Other carrier
	 * nodes can reassemble the fragments.
	 *
	 * Message may not be empty or null.
	 * @param
	 *	  message	 The message content defined by application
	 *
	 * @throws IllegalArgumentException illegal exception.
	 * @throws CarrierException  carrier exception.
	 */
	public void sendMessage(byte[] message) throws CarrierException {
		if (message == null || message.length == 0 || message.length > MAX_APP_MESSAGE_LEN)
			throw new IllegalArgumentException();

		if (!group_send_message(carrier, groupId, message))
			throw CarrierException.fromErrorCode(get_error_code());
	}

	/**
	 * Get group title.
	 *
	 * @return
	 * 		The title of the specified group
	 *
	 * @throws CarrierException  carrier exception.
	 */
	public String getTitle() throws CarrierException {
		String title = group_get_title(carrier, groupId);
		if (title == null)
			throw CarrierException.fromErrorCode(get_error_code());

		return title;
	}

	/**
	 * Set group title.
	 *
	 * @param
	 *	  title	   The title name to set(should be no longer than MAX_GROUP_TITLE_LEN)
	 *
	 * @throws IllegalArgumentException illegal exception.
	 * @throws CarrierException  carrier exception.
	 */
	public void setTitle(String title) throws CarrierException {
		if (title == null || title.length() == 0 || title.length() > MAX_GROUP_TITLE_LEN)
			throw new IllegalArgumentException();

		if (!group_set_title(carrier, groupId, title))
			throw CarrierException.fromErrorCode(get_error_code());
	}

	/**
	 * A class representing group peer information.
	 *
	 * Include the basic userid and it's nickname.
	 */
	static public class PeerInfo {
		private String userId;
		private String name;

		private PeerInfo(String name, String userId) {
			this.userId = userId;
			this.name = name;
		}

		public String getName() {
			return name;
		}

		public String getUserId() {
			return userId;
		}
	}

	/**
	 * Get group peer list.
	 *
	 * @return
	 * 		A list of all peers in the specified group
	 *
	 * @throws CarrierException  carrier exception.
	 */
	public List<PeerInfo> getPeers() throws CarrierException {
		List<PeerInfo> peers = new ArrayList<PeerInfo>();
		boolean result = group_get_peers(carrier, groupId, new GroupPeersIterator() {
			public boolean onIterated(PeerInfo peerInfo, Object context) {
				@SuppressWarnings({"unchecked"})
				List<PeerInfo> peers = (List<PeerInfo>)context;
				if (peerInfo != null)
					peers.add(peerInfo);
				return true;
			}
		}, peers);

		if (!result)
			throw CarrierException.fromErrorCode(get_error_code());

		return peers;
	}

	/**
	 * Get group peer information.
	 *
	 * @param
	 *	  peerId	  The target peerId to get it's information
	 *
	 * @return
	 * 		Information of the specified peer
	 *
	 * @throws IllegalArgumentException illegal exception.
	 * @throws CarrierException  carrier exception.
	 */
	public PeerInfo getPeer(String peerId) throws CarrierException {
		if (peerId == null || peerId.length() == 0)
			throw new IllegalArgumentException();

		PeerInfo peerInfo = group_get_peer(carrier, groupId, peerId);
		if (peerInfo == null)
			throw CarrierException.fromErrorCode(get_error_code());

		return peerInfo;
	}
}
